src/sudoku.hs

```diff
@@ -0,0 +1,136 @@
+module Sudoku where
+import GHC.Arr (fill)
+import Data.List
+
+-- | A representation of the values in a sudoku from 0-9 TODO: SHOULD BE RENAMED!!!!!
+data SudVal = One | Two | Three | Four | Five | Six | Seven | Eight | Nine
+    deriving (Eq, Enum, Bounded)
+
+instance Show SudVal where
+    show One = "1"
+    show Two = "2"
+    show Three = "3"
+    show Four = "4"
+    show Five = "5"
+    show Six = "6"
+    show Seven = "7"
+    show Eight = "8"
+    show Nine = "9"
+
+data Line = Horizontal | Vertical
+    deriving (Eq)
+
+instance Show Line where
+    show Horizontal = "-"
+    show Vertical = "|"
+
+data Note = Candidate SudVal | Line Line SudVal
+    deriving (Eq)
+
+instance Show Note where
+    show (Candidate x) = show x
+    show (Line l x) = show l ++ show x
+
+data Value
+    = Filled SudVal
+    | Note [Note]
+    | Empty
+    deriving (Eq)
+
+instance Show Value where
+    show (Filled x) = show x
+    show (Note xs) = show xs
+    show Empty = " "
+
+-- | A representation of a sudoku [Rows] 
+type Sudoku = [[Value]]
+
+-- | a position in a soduku (Row, Col)
+type Position = (Int, Int)
+
+-- | A secton is a collection of 9 values, either a row, column or block
+type Section =  [Value]
```

Comment:
Maybe better: a section is a collection of nine *positions*. Then later, given a concrete sudoku state, you can always extract the values from the positions.